qt/widgets/instrumentview/src/Shape2D.cpp

```diff
@@ -620,6 +623,497 @@ std::string Shape2DRing::saveToProject() const {
+             std::abs(BBox.bottomRight().y() - bRectBottomRight.y()) <
+                 epsilon) {
+
+    // top right corner is moving
+    computeScaling(BBox.topRight(), BBox.bottomLeft(),
+                   QPointF(bRectBottomRight.x(), bRectTopLeft.y()), 3);
+  }
+
+  // check if the shape has moved
+  if ((BBox.bottomRight().x() != bRectBottomRight.x() &&
+       BBox.topLeft().x() != bRectTopLeft.x() &&
+       std::abs((BBox.bottomRight().x() - bRectBottomRight.x()) -
+                (BBox.topLeft().x() - bRectTopLeft.x())) < epsilon) ||
+      (BBox.bottomRight().y() != bRectBottomRight.y() &&
+       BBox.topLeft().y() != bRectTopLeft.y() &&
+       std::abs((BBox.bottomRight().y() - bRectBottomRight.y()) -
+                (BBox.topLeft().y() - bRectTopLeft.y())) < epsilon)) {
+    // every corner has moved by the same distance -> the shape is being moved
+    qreal xDiff = bRectBottomRight.x() - BBox.bottomRight().x();
+    qreal yDiff = bRectBottomRight.y() - BBox.bottomRight().y();
+
+    m_center.setX(m_center.x() + xDiff);
+    m_center.setY(m_center.y() + yDiff);
+    resetBoundingRect();
+  }
+}
+
+void Shape2DSector::computeScaling(QPointF BBoxCorner,
+                                   QPointF BBoxOpposedCorner,
+                                   QPointF bRectCorner, int vertexIndex) {
+
+  // first we need to find the best projection of the new corner on the
+  // diagonal line of the rectangle, so its shape won't be modified, only
+  // scaled.
+  QPointF xProj, yProj, proj;
+  qreal xPos, yPos;
+  QVector2D slope;
+
+  slope = QVector2D(BBoxCorner - BBoxOpposedCorner);
+  xPos = (bRectCorner - BBoxCorner).x();
+  yPos = slope.y() * xPos / slope.x(); // TODO : check if non zero
+  xProj.setX(xPos);
+  xProj.setY(yPos);
+
+  yPos = (bRectCorner - BBoxCorner).y();
+  xPos = slope.x() * yPos / slope.y();
+
+  yProj.setX(xPos);
+  yProj.setY(yPos);
+
+  if (distanceBetween(xProj, QPointF(0, 0)) <
+      distanceBetween(yProj, QPointF(0, 0))) {
+    proj = xProj;
+  } else {
+    proj = yProj;
+  }
+  proj += BBoxCorner;
+  // then we need to adapt the shape to the new size
+  qreal ratio = distanceBetween(proj, BBoxOpposedCorner) /
+                distanceBetween(slope.toPointF(), QPointF(0, 0));
+
+  m_boundingRect.setVertex(vertexIndex, proj);
+
+  m_innerRadius *= ratio;
+  m_outerRadius *= ratio;
+  m_center.setX((m_center.x() - BBoxOpposedCorner.x()) * ratio +
+                BBoxOpposedCorner.x());
+  m_center.setY((m_center.y() - BBoxOpposedCorner.y()) * ratio +
+                BBoxOpposedCorner.y());
+}
+
+double distanceBetween(const QPointF &p0, const QPointF &p1) {
+  return sqrt(pow(p0.x() - p1.x(), 2) + pow(p0.y() - p1.y(), 2));
+}
+
+void Shape2DSector::resetBoundingRect() {
+  // based on the values of the different parameters defining a sector, compute
+  // the new bounding box
+  double x_min, x_max, y_min, y_max;
+  if ((m_startAngle <= M_PI / 2 && m_endAngle >= M_PI / 2) ||
```

Comment:
Are these conditions the same as in the `findArcBoundingBox`? If so can you make helper functions to centralize these checks?